File: mpp-proiect-java-culbec/MotorcycleContest/Persistence/src/test/java/RepositoryTests.java
import ConnectionUtils.DBUtils;
import Model.Participant;
import Model.User;
import Repository.*;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.security.crypto.bcrypt.BCrypt;

import java.io.FileReader;
import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.List;
import java.util.Optional;
import java.util.Properties;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class RepositoryTests {
    private final UserRepository userRepository;
    private final ParticipantRepository participantRepository;
    private final PreparedStatement insertUser;
    private final PreparedStatement deleteUser;
    private final PreparedStatement deleteParticipant;

    public RepositoryTests() throws RepositoryException, SQLException {
        Properties properties = new Properties();
        try {
            properties.load(new FileReader("DB.config"));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        DBUtils dbUtils = new DBUtils(properties);
        this.userRepository = new UserDBRepository(dbUtils);
        this.participantRepository = new ParticipantDBRepository(dbUtils);

        Connection connection = dbUtils.getConnection();
        this.insertUser = connection.prepareStatement("insert into users(uid, first_name, last_name, username, password) values(?, ?, ?, ?, ?)");
        this.deleteUser = connection.prepareStatement("delete from users where username = ?");

        this.deleteParticipant = connection.prepareStatement("delete from participants where first_name = ?");
    }

    private void beforeUser() throws SQLException {
        String test1Pass = BCrypt.hashpw("1234", BCrypt.gensalt());
        String test2Pass = BCrypt.hashpw("suzuki_power", BCrypt.gensalt());

        insertUser.setString(1, UUID.randomUUID().toString());
        insertUser.setString(2, "Test1F");
        insertUser.setString(3, "Test1L");
        insertUser.setString(4, "test1");
        insertUser.setString(5, test1Pass);

        insertUser.executeUpdate();

        insertUser.setString(1, UUID.randomUUID().toString());
        insertUser.setString(2, "Test2F");
        insertUser.setString(3, "Test2L");
        insertUser.setString(4, "test2");
        insertUser.setString(5, test2Pass);

        insertUser.executeUpdate();
    }

    private void afterUser() throws SQLException {
        deleteUser.setString(1, "test1");
        deleteUser.executeUpdate();

        deleteUser.setString(1, "test2");
        deleteUser.executeUpdate();
    }

    @Test
    @DisplayName("Model.User Repository - Repository Test")
    public void UserRepositoryTest() throws SQLException, RepositoryException {
        this.beforeUser();

        Optional<User> test1UserOptionalValid = userRepository.findUserByCredentials("test1", "1234");
        assertTrue(test1UserOptionalValid.isPresent(), "The user should be present.");
        assertEquals("Test1F", test1UserOptionalValid.get().getFirstName(), "The first name should be 'Test1F'.");
        assertEquals("Test1L", test1UserOptionalValid.get().getLastName(), "The last name should be 'Test1L'.");
        assertEquals("test1", test1UserOptionalValid.get().getUsername(), "The username should be 'test1'.");

        Optional<User> test1UserOptionalInvalid = userRepository.findUserByCredentials("test1", "12345");
        assertTrue(test1UserOptionalInvalid.isEmpty(), "The user should not be present.");

        Optional<User> test2UserOptionalValid = userRepository.findUserByCredentials("test2", "suzuki_power");
        assertTrue(test2UserOptionalValid.isPresent(), "The user should be present.");
        assertEquals("Test2F", test2UserOptionalValid.get().getFirstName(), "The first name should be 'Test2F'.");
        assertEquals("Test2L", test2UserOptionalValid.get().getLastName(), "The last name should be 'Test2L'.");
        assertEquals("test2", test2UserOptionalValid.get().getUsername(), "The username should be 'test2'.");

        this.afterUser();
    }

    private void afterParticipant() throws SQLException {
        deleteParticipant.setString(1, "Test1F");
        deleteParticipant.executeUpdate();

        deleteParticipant.setString(1, "Test2F");
        deleteParticipant.executeUpdate();
    }

    @Test
    @DisplayName("Model.Participant Repository - Repository Test")
    public void ParticipantRepositoryTest() throws SQLException, RepositoryException {
        Participant participant1 = new Participant.Builder()
                .setId(UUID.randomUUID())
                .setFirstName("Test1F")
                .setLastName("Test1L")
                .setTeam("Test1T")
                .setEngineCapacity(1000)
                .build();
        Participant participant2 = new Participant.Builder()
                .setId(UUID.randomUUID())
                .setFirstName("Test2F")
                .setLastName("Test2L")
                .setTeam("Test2T")
                .setEngineCapacity(2000)
                .build();

        Optional<Participant> participantOptional1 = participantRepository.save(participant1);
        assertTrue(participantOptional1.isPresent(), "The participant should not be present.");

        Optional<Participant> participantOptional2 = participantRepository.save(participant2);
        assertTrue(participantOptional2.isPresent(), "The participant should not be present.");

        Optional<Participant> existingParticipant = participantRepository.findParticipantByFields(participant1);
        assertTrue(existingParticipant.isPresent(), "The participant should be present.");

        Optional<Participant> participantByFields = participantRepository.findParticipantByFields(participant1);
        assertTrue(participantByFields.isPresent(), "The participant should be present.");
        assertEquals(participant1, participantByFields.get(), "The participant should be the same.");

        List<Participant> participantsByTeam = (List<Participant>) participantRepository.findParticipantsByTeam("Test1T");
        assertEquals(1, participantsByTeam.size(), "The number of participants should be 1.");

        this.afterParticipant();
    }
}

File: mpp-proiect-csharp-culbec/Tests/RepositoryTests.cs
using System.ComponentModel;
using System.Configuration;
using Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter;
using Model;
using Persistence.ConnectionUtils;
using Persistence.Repository;

namespace Tests;

public class RepositoryTests
{
    private static readonly RepositoryUserDb RepositoryUser;
    private static readonly RepositoryParticipantDb RepositoryParticipant;
    private static readonly SortedList<string, string> Properties = new();

    static RepositoryTests()
    {
        var connectionString = GetConnectionStringByName("MariaDBConnectionString");
        if (connectionString == null) throw new TestCanceledException("Cannot get configuration for the DB!");

        Console.WriteLine("ConnectionString for DB {0}", connectionString);
        Properties.Add("ConnectionString", connectionString);

        RepositoryUser = new RepositoryUserDb(Properties);
        RepositoryParticipant = new RepositoryParticipantDb(Properties);
    }

    [SetUp]
    [DisplayName("Repository Tests Setup")]
    public void Setup()
    {
        AddUsers(Properties);
    }

    [TearDown]
    [DisplayName("Repository Test Teardown")]
    public void TearDown()
    {
        RemoveUsers(Properties);
        RemoveParticipants(Properties);
    }

    [Test]
    [DisplayName("Repository Tests - User Repository Test")]
    public void UserRepositoryTest()
    {
        var test1UserValid = RepositoryUser.FindUserByCredentials("test1", "1234");
        Assert.That(test1UserValid, Is.Not.Null);
        Assert.Multiple(() =>
        {
            Assert.That(test1UserValid!.FirstName, Is.EqualTo("Test1F"));
            Assert.That(test1UserValid.LastName, Is.EqualTo("Test1L"));
            Assert.That(test1UserValid.Username, Is.EqualTo("test1"));
        });

        var test1UserInvalid = RepositoryUser.FindUserByCredentials("test1", "12345");
        Assert.That(test1UserInvalid, Is.Null);

        var test2UserValid = RepositoryUser.FindUserByCredentials("test2", "suzuki_power");
        Assert.That(test2UserValid, Is.Not.Null);
        Assert.Multiple(() =>
        {
            Assert.That(test2UserValid!.FirstName, Is.EqualTo("Test2F"));
            Assert.That(test2UserValid.LastName, Is.EqualTo("Test2L"));
            Assert.That(test2UserValid.Username, Is.EqualTo("test2"));
        });
    }

    [Test]
    [DisplayName("Repository Tests - Participant Repository Test")]
    public void ParticipantRepositoryTest()
    {
        var participant1 = Participant.NewBuilder()
            .SetId(Guid.NewGuid())
            .SetFirstName("Test1F")
            .SetLastName("Test1L")
            .SetTeam("Test1T")
            .SetEngineCapacity(1000)
            .Build();
        var participant2 = Participant.NewBuilder()
            .SetId(Guid.NewGuid())
            .SetFirstName("Test2F")
            .SetLastName("Test2L")
            .SetTeam("Test2T")
            .SetEngineCapacity(2000)
            .Build();

        var participant1Saved = RepositoryParticipant.Save(participant1);
        Assert.That(participant1Saved, Is.Null);

        var participant2Saved = RepositoryParticipant.Save(participant2);
        Assert.That(participant2Saved, Is.Null);

        participant1Saved = RepositoryParticipant.Save(participant1);
        Assert.That(participant1Saved, Is.Not.Null);

        var participant1ByFields = RepositoryParticipant.FindParticipantByFields(participant1);
        Assert.That(participant1ByFields, Is.EqualTo(participant1));

        var participantsByTeam = RepositoryParticipant.FindParticipantsByTeam("Test1T");
        Assert.That(participantsByTeam.Count(), Is.EqualTo(1));
    }

    private static void AddUsers(IDictionary<string, string> properties)
    {
        var test1Pass = BCrypt.Net.BCrypt.HashPassword("1234");
        var test2Pass = BCrypt.Net.BCrypt.HashPassword("suzuki_power");

        using var connection = DBUtils.GetConnection(properties);
        using var command = connection.CreateCommand();
        command.CommandText =
            "insert into users(first_name, last_name, username, password) values " +
            "(@firstName1, @lastName1, @username1, @password1), " +
            "(@firstName2, @lastName2, @username2, @password2)";

        var paramFirstName1 = command.CreateParameter();
        paramFirstName1.ParameterName = "@firstName1";
        paramFirstName1.Value = "Test1F";
        command.Parameters.Add(paramFirstName1);

        var paramLastName1 = command.CreateParameter();
        paramLastName1.ParameterName = "@lastName1";
        paramLastName1.Value = "Test1L";
        command.Parameters.Add(paramLastName1);

        var paramUsername1 = command.CreateParameter();
        paramUsername1.ParameterName = "@username1";
        paramUsername1.Value = "test1";
        command.Parameters.Add(paramUsername1);

        var paramPassword1 = command.CreateParameter();
        paramPassword1.ParameterName = "@password1";
        paramPassword1.Value = test1Pass;
        command.Parameters.Add(paramPassword1);

        var paramFirstName2 = command.CreateParameter();
        paramFirstName2.ParameterName = "@firstName2";
        paramFirstName2.Value = "Test2F";
        command.Parameters.Add(paramFirstName2);

        var paramLastName2 = command.CreateParameter();
        paramLastName2.ParameterName = "@lastName2";
        paramLastName2.Value = "Test2L";
        command.Parameters.Add(paramLastName2);

        var paramUsername2 = command.CreateParameter();
        paramUsername2.ParameterName = "@username2";
        paramUsername2.Value = "test2";
        command.Parameters.Add(paramUsername2);

        var paramPassword2 = command.CreateParameter();
        paramPassword2.ParameterName = "@password2";
        paramPassword2.Value = test2Pass;
        command.Parameters.Add(paramPassword2);

        command.ExecuteNonQuery();
    }

    private static void RemoveUsers(IDictionary<string, string> properties)
    {
        using var connection = DBUtils.GetConnection(properties);
        using var command = connection.CreateCommand();

        command.CommandText = "delete from users where username = @username1 or username = @username2";

        var paramUsername1 = command.CreateParameter();
        paramUsername1.ParameterName = "@username1";
        paramUsername1.Value = "test1";
        command.Parameters.Add(paramUsername1);

        var paramUsername2 = command.CreateParameter();
        paramUsername2.ParameterName = "@username2";
        paramUsername2.Value = "test2";
        command.Parameters.Add(paramUsername2);

        command.ExecuteNonQuery();
    }

    private static void RemoveParticipants(IDictionary<string, string> properties)
    {
        using var connection = DBUtils.GetConnection(properties);
        using var command = connection.CreateCommand();

        command.CommandText = "delete from participants where first_name = @firstName1 or first_name = @firstName2";

        var paramFirstName1 = command.CreateParameter();
        paramFirstName1.ParameterName = "@firstName1";
        paramFirstName1.Value = "Test1F";
        command.Parameters.Add(paramFirstName1);

        var paramFirstName2 = command.CreateParameter();
        paramFirstName2.ParameterName = "@firstName2";
        paramFirstName2.Value = "Test2F";
        command.Parameters.Add(paramFirstName2);

        command.ExecuteNonQuery();
    }

    private static string? GetConnectionStringByName(string name)
    {
        var appConfigPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "app.config");
        var fileMap = new ExeConfigurationFileMap { ExeConfigFilename = appConfigPath };
        var config = ConfigurationManager.OpenMappedExeConfiguration(fileMap, ConfigurationUserLevel.None);
        var settings = config.ConnectionStrings.ConnectionStrings[name];
        return settings?.ConnectionString;
    }
}